compiler/packages/babel-plugin-react-compiler/src/HIR/PruneNonEscapingScopesHIR.ts

```diff
@@ -0,0 +1,957 @@
+             * be memoized.
+             */
+            if (signature && signature.noAlias === true) {
+              continue;
+            }
+            for (const operand of instr.value.args) {
+              const place = operand.kind === 'Spread' ? operand.place : operand;
+              changed ||= !state.escapingValues.has(place.identifier.id);
+              state.escapingValues.add(place.identifier.id);
+            }
+          }
+        }
+      }
+
+      if (block.terminal.kind === 'return') {
+        changed ||= !state.escapingValues.has(
+          block.terminal.value.identifier.id,
+        );
+        state.escapingValues.add(block.terminal.value.identifier.id);
+      }
+    }
+  } while (changed);
+
+  CompilerError.invariant(undefinedIdentifiers == null, {
+    reason: 'Expected all identifiers to be initialized',
+    description: `No definition for ${undefinedIdentifiers?.().map(p => p.identifier.id)}`,
+    loc: fn.loc,
+    suggestions: null,
+  });
+
+  function visitAliasing(aliasing: Aliasing, instrId: InstructionId): boolean {
+    let changed = false;
+    // Associate all the rvalues with the instruction's scope if it has one
+    for (const operand of aliasing.rvalues) {
+      const operandId =
+        state.definitions.get(operand.identifier.id) ?? operand.identifier.id;
+      changed = state.visitOperand(instrId, operand, operandId) || changed;
+    }
+
+    // Add the operands as dependencies of all lvalues.
+    for (const {place: lvalue, level} of aliasing.lvalues) {
+      const lvalueId =
+        state.definitions.get(lvalue.identifier.id) ?? lvalue.identifier.id;
+      let node = state.identifiers.get(lvalueId);
+      if (node === undefined) {
+        node = {
+          level: MemoizationLevel.Never,
+          memoized: false,
+          dependencies: new Set(),
+          scopes: new Set(),
+          seen: false,
+        };
+        state.identifiers.set(lvalueId, node);
+        changed = true;
+      }
+      const newLevel = joinAliases(node.level, level);
+      if (newLevel !== node.level) {
+        node.level = newLevel;
+        changed = true;
+      }
+      /*
+       * This looks like NxM iterations but in practice all instructions with multiple
+       * lvalues have only a single rvalue
+       */
+      for (const operand of aliasing.rvalues) {
+        const operandId =
+          state.definitions.get(operand.identifier.id) ?? operand.identifier.id;
+        if (operandId === lvalueId) {
+          continue;
+        }
+        changed ||= !node.dependencies.has(operandId);
+        node.dependencies.add(operandId);
+      }
+
+      changed = state.visitOperand(instrId, lvalue, lvalueId) || changed;
+    }
+
+    return changed;
+  }
+}
+
+function pruneScopes(fn: HIRFunction, state: Set<IdentifierId>): void {
+  const prunedScopes = new Set<ScopeId>();
+  const reassignments = new Map<IdentifierId, Set<Identifier>>();
```

Comment:
the change to use IdentifierId instead of DeclarationId makes sense for most cases, but for reassignments it seems a bit suspect because the id value may not line up btw the reassignment and the FinishMemoize instruction.

to me what this points to is that we may not need this logic anymore, would be good to check if we can remove it (or at least add a TODO so we can come back to this)